vortex-array/src/array/bool/mod.rs

```diff
@@ -105,24 +109,22 @@ impl BoolArray {
             .into_inner()
             .bit_slice(buffer_byte_offset, buffer_len);
 
-        Ok(Self {
+        Self {
             typed: TypedArray::try_from_parts(
                 DType::Bool(validity.nullability()),
                 buffer_len,
                 BoolMetadata {
-                    validity: validity.to_metadata(buffer_len)?,
+                    validity: validity
+                        .to_metadata(buffer_len)
+                        .vortex_expect("Validity array has different length"),
```

Comment:
you should revert this `try_new` -> `new` part... the constructor is intrinsically fallible, we only want to panic on programmer error

Reply:
This is programming error. (We need to have a broader discussion around vortex errors, IMO they go too far.)

The only way this constructor fails is if someone has written some bad Rust. This is the same as Arrow's StructArray::new that panics if fields are different lengths.

Errors should be returned when runtime user input can be bad, which should be caught when validating the contents of ArrayData (a function that doesn't actually exist). It doesn't happen on this code path.

Reply:
I've compromised for you. There's a `BoolArray::new(buffer, nullability)` since the validity is either AllValid or NonNullable. Then I've kept the `try_new(buffer, validity)` where the validity length may not match.